Change rustdoc-scrape-examples to be a target-level configuration

willcrichton opened the pull request:
This PR addresses issues raised in rust-lang/cargo#9525. Specifically:
1. It enables examples to be scraped from `#[test]` functions, by passing additional flags to Rustdoc to ensure that these functions aren't ignored by rustc. 
2. It moves the `arg` from `-Z rustdoc-scrape-examples={arg}` into a target-level configuration that can be added to Cargo.toml. 

The added test `scrape_examples_configure_target` shows a concrete example. In short, examples will be default scraped from Example and Lib targets. Then the user can enable or disable scraping like so:

```toml
[lib]
doc-scrape-examples = false

[[test]]
name = "my_test"
doc-scrape-examples = true
```

camelid commented:
Somehow I feel like it'd be better to structure the configuration as something like:

```
[doc]
scrape-examples-targets = ["lib", "test"]
```

rather than putting `doc-scrape-example` in each section. Just a gut reaction though.

willcrichton commented:
> Somehow I feel like it'd be better to structure the configuration as something like:
> 
> ```
> [doc]
> scrape-examples-targets = ["lib", "test"]
> ```
> 
> rather than putting `doc-scrape-examples` in each target section. Just a gut reaction though.

I think the issue with this is (a) it's not consistent with existing ways to configure targets, and (b) it doesn't allow target configuration at a higher granularity. For instance, the proposed method allows you to individually enable/disable specific examples and tests, while a list of target strings can't do so.

alexcrichton commented:
This all looks pretty reasonable to me, but I'd like to confirm my understanding. The intention is that once stabilized this is enabled for all projects by default? So all projects with libraries and examples, when using `cargo doc`, will see examples for methods and such drawn from libraries examples. Packages have fine-grained control at the per-target level what's used as candidates for examples and what isn't (with the default being only the lib targets and examples)

If that's the case I think it might be good to flesh out the failing tests for now. I think there may need to be some trickery around not running doc-scrape with stable Rust as-is today since Cargo tests its CI with all three channels of Rust. Otherwise though that seems pretty reasonable to me and the PR otherwise looks pretty good.

>What is the right set of default targets to search for examples?

Personally the choice made here I think is fine. This seems like something we can alter over time if necessary as well.

> Is there a different granularity at which example scraping should also be configurable?
 
At least from Cargo's perspective this probably won't have much impact, but I don't have much of an opinion on this otherwise. 


----

As a random bikeshedding point seeing `doc-scrape-examples = false` in isolation feels a little surprising, so I might suggest something like `doc-example-candidate = false` or something like that.

GuillaumeGomez commented:
I thought this feature would be disabled by default?

willcrichton commented:
> I thought this feature would be disabled by default?

My understanding is that it would be disabled by default while unstable, and then enabled by default once stabilized. The vast majority of people will not use this feature if it is hidden behind a flag.

GuillaumeGomez commented:
> My understanding is that it would be disabled by default while unstable, and then enabled by default once stabilized. The vast majority of people will not use this feature if it is hidden behind a flag.

Sorry, I definitely didn't understand it like this. In this case, we'll need to update the display quite a lot because it takes a lot of space by default. What about collapsing the section by default with a title like "see usage in examples" or something among the line? If you prefer we can discuss it on zulip.

willcrichton commented:
The discussion about scraping-by-default-or-not has been pulled onto Zulip: https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Enabling.20scrape.20examples.20by.20default

bors commented:
:umbrella: The latest upstream changes (presumably #9992) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lqd commented:
@willcrichton The `.rustfmt.toml` change and new `log` file, seem unintentional.

willcrichton commented:
Ah thanks @lqd , got caught up in the rebase. I'll squash everything in a bit.

bors commented:
:umbrella: The latest upstream changes (presumably #10533) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

willcrichton commented:
This PR is ready for review, although it needs a new Cargo reviewer since Alex has stepped back.

As a reminder, this PR will enable the scrape examples feature by default, although only when run with a nightly rustc. Now that rust-lang/rust#93217 has landed, I think this feature is ready to be rolled out at least on docs.rs. Rustdoc maintainers, please let me know if you have any remaining concerns about this feature. 

cc @jsha @camelid @GuillaumeGomez

@rustbot ready

willcrichton commented:
It seems like only @ehuss is on [reviewer rotation](https://github.com/rust-lang/highfive/blob/master/highfive/configs/rust-lang/cargo.json) for Cargo right now, so I will add him. Sorry for all the extra work Eric :(

r? @ehuss

ehuss commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -370,8 +372,8 @@ pub fn create_bcx<'a, 'cfg>(
     let all_packages = &Packages::All;
-    let rustdoc_scrape_examples = &config.cli_unstable().rustdoc_scrape_examples;
-    let need_reverse_dependencies = rustdoc_scrape_examples.is_some();
+    let need_reverse_dependencies =
+        matches!(build_config.mode, CompileMode::Doc { .. }) && config.nightly_features_allowed;
```
There are a lot of local variables in this function, and I'm having a hard time keeping them straight. (spec, specs, full_specs, resolve_specs, need_reverse_dependencies).

Might I suggest a few things to maybe clarify it a little? Maybe something like:

```rust
let needs_scrape_examples =
    matches!(build_config.mode, CompileMode::Doc { .. }) && config.nightly_features_allowed;
let requested_specs = spec.to_package_id_specs(ws)?;
let specs_to_resolve = if needs_scrape_examples {
    // When scraping examples, this neads All because …
    Packages::All.to_package_id_specs(ws)?
} else {
    requested_specs.clone()
};
```

Where this:
* `need_reverse_dependencies` → `needs_scrape_examples`
* `resolve_specs` → `specs_to_resolve`
* Remove `full_specs` and `specs`
* Add `requested_specs` for the specs requested by the user

Also, I'm a bit surprised that scraping forces scraping from the entire workspace. That seems like it could cause a lot of problems. Am I understanding what this is doing correctly?

willcrichton replied:
I tried to clean up the code roughly in line with your suggestion. Let me know if you think it's clearer.

And yes, scraping does by default look through all the packages in your workspace for examples. What kind of problems are you anticipating?

ehuss replied:
I'm pretty sure that isn't something we can do by default. Workspaces will have packages that don't compile, or only build on certain targets, or have very expensive build scripts, or any number of requirements. Some workspaces may be huge, but the user only wants to document one crate. There are some users who have monolithic workspaces across entire orgs.

I think if the user wants to document the entire workspace, they can specify `--workspace` or multiple `-p` flags.

willcrichton replied:
Ok, that makes sense. I changed it to just use the package set requested by the user, and we can eventually document that users should do `--all` or whatever if they want to scrape all possible examples.

ehuss commented:
This seems to be a very unusual way of stabilizing something. Can you clarify exactly how this is intended to work? I'm skeptical that our setup can handle stabilizing in this fashion, so I'm a little concerned. Is there an intent to stabilize the rustdoc side? Usually that is done before the cargo side.

I'm also a little surprised to see this moving for stabilization so quickly. I don't have much of an impression that it has gotten much testing. I also haven't seen much discussion from the rustdoc side.

willcrichton commented on src/cargo/core/manifest.rs:
```diff
@@ -220,6 +220,7 @@ struct TargetInner {
     for_host: bool,
     proc_macro: bool,
     edition: Edition,
+    doc_scrape_examples: Option<bool>,
```
Sure, I implemented your suggested strategy in the most recent commit.

willcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -1203,6 +1205,8 @@ impl<'cfg> DrainState<'cfg> {
                 } else if unit.mode.is_doc_test() {
                     // Skip doc test.
+                } else if unit.mode.is_doc_scrape() {
+                    config.shell().status("Scraping", &unit.pkg)?;
```
I implemented the strategy you suggested. But if you don't think it's worth it to notify the user about scraping, I can also remove the notification.

ehuss replied:
This still doesn't seem to be working as I would expect.

In a project with `src/lib.rs` and `examples/foo.rs`, `cargo doc` doesn't mention "Scraping". I think the filter:

```rust
self.compiled.contains(&unit.pkg.package_id()) && !unit.mode.is_doc()
```

would need to be:

```rust
self.compiled.contains(&unit.pkg.package_id()) && !unit.mode.is_doc() && !unit.mode.is_doc_scrape()
```

As for whether or not to do it at all, I'm on the fence. We have felt that cargo is already a little too verbose (eventually we will likely get rid of all the status messages during a build).

If it were me, I would probably not include it.

ehuss replied:
I just realized that it does need to print *something*, otherwise it will sometimes say "Compiling" which probably isn't good.  I think with the above adjustment, it should probably be fine.

willcrichton replied:
Ah ok I see. I've fixed it with what you suggested, and it consistently indicates when scraping.

willcrichton commented on src/cargo/core/compiler/unit.rs:
```diff
@@ -101,7 +101,7 @@ impl UnitInner {
 impl Unit {
     pub fn buildkey(&self) -> String {
-        format!("{}-{}", self.pkg.name(), short_hash(self))
+        format!("{}-{}", self.pkg.name(), short_hash(&*self.inner))
```
This was to ensure that the buildkey was stable across sessions, but this fact is no longer needed.

willcrichton commented on src/cargo/core/features.rs:
```diff
@@ -887,15 +884,6 @@ impl CliUnstable {
             "namespaced-features" => stabilized_warn(k, "1.60", STABILISED_NAMESPACED_FEATURES),
             "weak-dep-features" => stabilized_warn(k, "1.60", STABILIZED_WEAK_DEP_FEATURES),
             "credential-process" => self.credential_process = parse_empty(k, v)?,
-            "rustdoc-scrape-examples" => {
```
Good idea, I added that in the latest commit.

willcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -1595,9 +1579,12 @@ fn rebuild_unit_graph_shared(
     let new_scrape_units = scrape_units
         .iter()
-        .map(|unit| memo.get(unit).unwrap().clone())
+        .filter_map(|unit| memo.get(unit).cloned())
```
I think the recent bugfix PRs addressed this issue, so I reverted this change.

willcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -1625,7 +1612,13 @@ fn traverse_and_share(
+            // be stable across Cargo sessions for Doscrape units to be memoized, so we
+            // take care to hash the contents of the unit here.
+            let inner: &UnitInner = &*new_dep_unit;
+            inner.hash(&mut dep_hash);
```
Ok yes, I think metadata is actually what I needed. So I changed this feature to use `metadata` instead of `buildkey`, and I also added a test to ensure that scraping is properly memoized. I reverted all changed to unit hashing, since they're no longer necessary.

willcrichton commented:
My goal here is to provide a bigger roll out for this feature before fully stabilizing it. There's been a lot of back-and-forth on the UI, but those discussions were finally settled (for now at least) in rust-lang/rust#93217. Essentially this change will activate the scrape examples feature by default on docs.rs, which uses nightly. More users will encounter the UI, and we we will be able to identify any remaining Cargo issues from broken doc builds. Once users are satisfied with the UI and all Cargo issues are settled, then we can finish stabilizing the feature.

Sorry, it may be misleading to say this PR "prepares the feature for stabilization". I mean preparing in a looser sense, as in "is the biggest remaining step before stabilization"

ehuss commented:
This does seem to actually stabilize it, though, in the sense that there is no longer a `-Z` flag to gate it. Cargo specifically has a system to make things like this require an explicit opt-in before it is stabilized. Users should be able to use the nightly channel the same as the stable channel.

I might suggest a different way to try to get more exposure. Would it be possible to just enable it on docs.rs by default? Have you talked to them about it?

Also, another tactic is to do a crater run with it enabled, just to see what basic problems might arise.  I would do that before docs.rs.  (I would also make sure the issues raised in this PR are fixed first.)

willcrichton commented:
Ok that makes sense. Since this isn't stable yet, I restored the `-Zrustdoc-scrape-examples` flag. But now instead of passing `=all` or `=examples`, that configuration is still done at the target level in Cargo.toml. I will also do a crater run this weekend to ferret out any issues. I'll check with the docs.rs folks and see if they're willing to add the flag.

willcrichton commented:
Update: I did a crater run of the scrape-examples feature as it exists on nightly on the top-10,000 crates (after implementing rust-lang/crater#626). In total there were 131 regressions. I could not find any issues directly in the implementation of scrape-examples. Instead, regressions had one of three causes:

1. The project's examples did not type-check.
2. The project itself did not type-check (e.g. `cargo check` would fail while `cargo doc` would work).
3. Dependencies of the examples did not type-check.

Aside: I'm actually a bit surprised that `cargo publish` includes examples by default, but doesn't check if they compile. Is that intended behavior? Ensuring that examples work at publish-time would avoid the first issue.

So the question then is whether these regressions matter. If it's not important that `cargo doc` should still work in these cases, then I think the feature is ready for deployment on docs.rs, especially once this PR lands. But if it _is_ important, then I think the only fix would be to somehow mark the scrape units in Cargo as "fallible" or "optional", like if they fail, the build should still continue. I'm not sure if Cargo has such a concept, or how easy it would be to implement. 

@ehuss what are your thoughts on this?

willcrichton commented:
After some discussion with @jyn514 on Zulip, it's clear that a `Docscrape` unit should be able to fail without killing the build. Some crates (but not most!) will rely on the fact that `cargo doc` will work while `cargo check` will not, and that is fundamental to the design of rustdoc.

So I think what I'm going to do is keep this PR scoped to its original goal, of changing the target configuration for the scrape-examples feature (plus the couple bugfixes that have emerged in review). Then I'm going to put up a second PR that adds the ability to mark units as "ok to fail", which will be like a more granular version of `BuildConfig::keep_going`, and mark all Docscrape units as such. So @ehuss let me know if you have any more comments on this PR as is.

bors commented:
:umbrella: The latest upstream changes (presumably #10868) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

willcrichton commented:
@ehuss I just implemented the strategy you described [here](https://github.com/rust-lang/cargo/pull/10596#issuecomment-1186773249), and added a set of corresponding unit tests. Let me know if you think this captures your intent.

One implementation note: I'm using `state.emit_diag("warning", ...)` to emit the warning when silenced scraped units fail. I'm not sure if that's the correct API to use.

ehuss commented:
Can we move all of the docscrape tests to a separate module?  Having `doc.rs` climb past 3,000 lines is making it a bit too heavy-weight, and it can be convenient to keep new features in their own little scope. I'll leave it up to you if it would be easier to do before or after this PR.

ehuss commented:
There doesn't seem to be anything that will enabling scraping of targets with `doc-scrape-examples=true` set (or to disable it if `doc-scrape-examples=false`). 

I see that may be difficult to implement because `has_dev_units` needs to be determined before running the resolver, but it doesn't know what is necessarily being built before running the resolver. Perhaps one workaround is to call `ws.members_with_features(…)` and check if any of those packages have example targets with `doc-scrape-examples=true`, and if so enable dev-dependencies, and change the `CompileFilter` to include those targets.

Conversely, I imagine there should be some filter somewhere to prevent targets with `doc-scrape-examples=false` from being implicitly included.

willcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -686,6 +673,56 @@ pub fn create_bcx<'a, 'cfg>(
+    has_dev_units: HasDevUnits,
+    filter: &CompileFilter,
+) -> CompileFilter {
+    let no_pkg_has_dev_deps = pkg_set.packages().all(|pkg| {
```
Ah yeah that's right. Using `to_builds` seems reasonable, I changed this function to use that set instead of `PackageSet`.

willcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -640,13 +640,17 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
     if unit.mode.is_doc_scrape() {
         debug_assert!(cx.bcx.scrape_units.contains(unit));
 
+        if unit.target.is_test() {
+            rustdoc.arg("--scrape-tests");
+        }
```
It isn't possible in that you can't pass `--tests` to `cargo doc`. But re: later in your review, if we make it so `[[tests]] doc-scrape-examples = true` will proactively include the test target, then this could arise. So it's a bit of future proofing.

willcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -708,22 +740,42 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
-                        &mut output_options,
-                    )
+                    if silence_errors {
+                        Ok(())
```
Good point, I think just setting `show_diagnostics` is the simpler approach. That at least passes my simple test.

willcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -708,22 +740,42 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
+            if silence_errors && output_options.show_diagnostics {
+                let diag = format!(
+                    "\
+failed to scan example \"{target_name}\" in package `{name}` for example code usage
```
Done.

willcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -708,22 +740,42 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
-            .with_context(|| format!("could not document `{}`", name))?;
+            .with_context(|| format!("could not document `{}`", name));
+
+        if let Err(e) = result {
```
Rustdoc allows functions without valid bodies. If you change it to something like `fn main();` then Rustdoc will return a nonzero exit code.

willcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -708,22 +740,42 @@ fn rustdoc(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Work> {
+    If this example should not be scanned, consider adding `doc-scrape-examples = false` to the `[[example]]` definition in {}",
+                    relative_manifest_path.display()
+                );
+                state.emit_diag("warning".into(), diag)?;
```
Ok, I implemented the strategy you suggested.

One issue though is I'm not sure how this special pathway should interact with the deduplication functionality. It seems like deduplication is handled both within the child process and the parent process?

willcrichton commented:
> Can we move all of the docscrape tests to a separate module? Having doc.rs climb past 3,000 lines is making it a bit too heavy-weight, and it can be convenient to keep new features in their own little scope. I'll leave it up to you if it would be easier to do before or after this PR.

Done, moved to a new `docscrape.rs`.

> There doesn't seem to be anything that will enabling scraping of targets with doc-scrape-examples=true set (or to disable it if doc-scrape-examples=false).
> I see that may be difficult to implement because has_dev_units needs to be determined before running the resolver, but it doesn't know what is necessarily being built before running the resolver. Perhaps one workaround is to call ws.members_with_features(…) and check if any of those packages have example targets with doc-scrape-examples=true, and if so enable dev-dependencies, and change the CompileFilter to include those targets.

I just added that here, based on the strategy you recommended:

https://github.com/rust-lang/cargo/pull/10343/files#diff-98da1b66b532e50e9bca971b453ee7de96b17436de1d7ef3824c9888d55be9bbR377-R395

> Conversely, I imagine there should be some filter somewhere to prevent targets with doc-scrape-examples=false from being implicitly included.

That is the purpose of this check after running `generate_targets`:

https://github.com/rust-lang/cargo/pull/10343/files#diff-98da1b66b532e50e9bca971b453ee7de96b17436de1d7ef3824c9888d55be9bbR535-R543

willcrichton commented:
@rustbot label +S-waiting-on-author -S-waiting-on-review

bors commented:
:umbrella: The latest upstream changes (presumably #9892) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

willcrichton commented:
I did another Crater run on the top 10k packages with the latest scrape examples implementation. I realized that the current strategy does not handle the following case:

1. A crate's library *documents* but does not *compile*.
2. The crate has an example and no dev-dependencies.
3. Cargo tries to opportunistically scan the example.
4. To scan the example, Cargo needs to check the example's dependencies, which includes the crate library.
5. Cargo compiles the crate library and fails. The lib-check unit is *not* opportunistic, so its failure crashes the entire build.

I can only think of a few solutions to this:
1.  **General solution:** essentially what I proposed in #10596. If *any transitive dependency* of a fallible (or opportunistic or what have you) unit fails, then that has to kill the build *up to the fallible dependency*. This will solve the problem robustly, but adds a lot of new concepts and implementation complexity.
2. **Specific solution:** if a lib-check unit fails AND it's a dependency of a scrape unit, then don't crash the build. This solves the specific problem, but adds yet another edge case to Cargo.
3. **Scrape examples becomes opt-in:** we give up on trying to make any kind of default behavior for scrape examples, and users have to opt-in by passing some kind of flag or Cargo.toml parameter. This is guaranteed backward-compatible, but would cause the feature to be used far less.

@ehuss What do you think strikes the right balance between feature impact and Cargo tech debt? Or do you see another possible solution?

jyn514 commented:
@willcrichton just fyi - I would recommend trying to land the changes you have so far before trying to make more. This PR is already quite big and it's hard to skim the comments, having a separate issue for the problem you found would be helpful I think.

willcrichton commented:
That's sensible, I don't want to expand the scope of this PR too much. I'm happy to have this PR be merged as-is if we want to pull this discussion elsewhere.

bors commented:
:umbrella: The latest upstream changes (presumably #10929) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

willcrichton commented:
@ehuss this PR is ready for review whenever you have the time. I'll fix the issue described in https://github.com/rust-lang/cargo/pull/10343#issuecomment-1203170975 in a later PR.

willcrichton commented:
@ehuss do you still have the cycles to review this PR? If not that's totally fine, I can ask epage or weihanglo to take over.

bors commented:
:umbrella: The latest upstream changes (presumably #11207) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

willcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -353,6 +354,12 @@ enum Message {
     },
+    // This is distinct from Diagnostic because it gets emitted through
+    // a different Shell pathway
+    Warning {
```
I have added an explanation, let met know if it makes sense.

willcrichton commented on src/cargo/core/compiler/rustdoc.rs:
```diff
@@ -190,3 +190,20 @@ pub fn add_root_urls(
 }
+
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Debug, Copy)]
+pub enum RustdocScrapeExamples {
```
I have added a doc comment, let me know if it makes sense.

willcrichton commented on src/cargo/core/manifest.rs:
```diff
@@ -21,6 +21,8 @@ use crate::util::interning::InternedString;
 use crate::util::toml::{TomlManifest, TomlProfiles};
 use crate::util::{short_hash, Config, Filesystem};
 
+use super::compiler::rustdoc::RustdocScrapeExamples;
```
Done.

willcrichton commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -77,6 +77,8 @@ pub struct Context<'a, 'cfg> {
     /// See Context::find_metadata_units for more details.
     pub metadata_for_doc_units: HashMap<Unit, Metadata>,
+
+    pub failed_scrape_units: Arc<Mutex<HashSet<Metadata>>>,
```
I have added a doc comment, let me know if it makes sense.

willcrichton commented on src/cargo/ops/cargo_compile/mod.rs:
```diff
@@ -565,6 +567,56 @@ pub fn create_bcx<'a, 'cfg>(
 
+/// Generate a CompileFilter that represents the maximal set of targets that should be
+/// considered for scraping. Should be a subset of the CLI-provided CompileFilter.
+fn filter_for_scrape_units(
```
Done.

willcrichton commented on tests/testsuite/package_features.rs:
```diff
@@ -468,9 +468,7 @@ fn non_member() {
-        .with_stderr(
-            "[UPDATING][..]\n[ERROR] cannot specify features for packages outside of workspace",
-        )
+        .with_stderr("[ERROR] cannot specify features for packages outside of workspace")
```
This change causes the error to happen earlier. https://github.com/rust-lang/cargo/pull/10343/files#diff-9b38573ab038c139588eedf8c9e475e5ff047deb3b8335ca4fa94642e9b31fe6R238-R267

willcrichton commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -426,6 +433,14 @@ impl<'a, 'cfg> JobState<'a, 'cfg> {
         Ok(())
     }
 
+    pub fn warning(&self, warning: String) -> CargoResult<()> {
```
I added a reference back to the explanation on the relevant enum.

willcrichton commented on tests/testsuite/docscrape.rs:
```diff
@@ -0,0 +1,518 @@
+}
+
+#[cargo_test(nightly, reason = "rustdoc scrape examples flags are unstable")]
+fn configure_profile() {
```
This is from #10500, I changed the name to reflect that.

willcrichton commented on tests/testsuite/docscrape.rs:
```diff
@@ -0,0 +1,518 @@
+}
+
+#[cargo_test(nightly, reason = "rustdoc scrape examples flags are unstable")]
+fn no_scrape_with_dev_deps() {
```
I added an explanation, let me know if it makes sense.

willcrichton commented on src/cargo/ops/cargo_compile/mod.rs:
```diff
@@ -565,6 +567,56 @@ pub fn create_bcx<'a, 'cfg>(
+        FilterRule::All
+    } else {
+        FilterRule::none()
+    };
```
I added a lengthy explanation (that also relates to the `no_scrape_with_dev_deps` test). Let me know if it makes sense.

weihanglo replied:
I love your writing. Thank you for providing such a good explanation!

willcrichton commented:
@weihanglo thanks for the review. I've either fixed or responded to all of your comments.

weihanglo commented:
This would fix https://github.com/rust-lang/cargo/issues/10876 as well, right?

willcrichton commented:
> This would fix #10876 as well, right?

That's correct. I just added the suggested change to the `docscrape::basic` test to confirm this.

weihanglo reviewed (approved):
Thanks for the update! I believe this pull request is effectively an implementation based on ehuss's comment <https://github.com/rust-lang/cargo/pull/10596#issuecomment-1186773249>, plus fingerprint support for scrape units. Totally agree it is the best effort to roll out this feature by default with minimal intrusion. 

Since this is still unstable, I am going to merge it without an FCP. If anyone thinks Cargo shouldn't scrape examples under those conditions, we can revisit it and change the logic of compile filters.

bors commented:
:pushpin: Commit 183425f0bce8a7d32d8ace09552bd1c85559fb9f has been approved by `weihanglo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=weihanglo 183425f0bce8a7d32d8ace09552bd1c85559fb9f -->
<!-- homu: {"type":"Approved","sha":"183425f0bce8a7d32d8ace09552bd1c85559fb9f","approver":"weihanglo","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 183425f0bce8a7d32d8ace09552bd1c85559fb9f with merge 99d3c71cbe1dd6d0d30a1b93aa08096255c7f3bb...
<!-- homu: {"type":"BuildStarted","head_sha":"183425f0bce8a7d32d8ace09552bd1c85559fb9f","merge_sha":"99d3c71cbe1dd6d0d30a1b93aa08096255c7f3bb"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3540935149/jobs/5945136171)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/cargo/actions/runs/3540935149/jobs/5945136171","builder_name":"checks-actions"} -->

willcrichton commented:
Seems like the test `lto::test_profile` failed, but I don't think anything in this PR would have affected it. Any reason to believe it's flaky?

weihanglo commented:
I am not sure if it is related to https://github.com/rust-lang/rust/issues/98302, but [it can be reproduced](https://github.com/weihanglo/cargo/actions/runs/3541971990/jobs/5946901773) so seems not a flaky test?

Anyway, let's try again.

@bors retry

bors commented:
:hourglass: Testing commit 183425f0bce8a7d32d8ace09552bd1c85559fb9f with merge 3009cbbb1c61364f455cbd6f340fde4ff91f99f4...
<!-- homu: {"type":"BuildStarted","head_sha":"183425f0bce8a7d32d8ace09552bd1c85559fb9f","merge_sha":"3009cbbb1c61364f455cbd6f340fde4ff91f99f4"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3542312689/jobs/5948119704)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/cargo/actions/runs/3542312689/jobs/5948119704","builder_name":"checks-actions"} -->

weihanglo commented:
@bors retry

bors commented:
:hourglass: Testing commit 183425f0bce8a7d32d8ace09552bd1c85559fb9f with merge de56c1251bfe8f864112bc6cd9e9d73fee71fac4...
<!-- homu: {"type":"BuildStarted","head_sha":"183425f0bce8a7d32d8ace09552bd1c85559fb9f","merge_sha":"de56c1251bfe8f864112bc6cd9e9d73fee71fac4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3545958152/jobs/5954969068)
Approved by: weihanglo
Pushing de56c1251bfe8f864112bc6cd9e9d73fee71fac4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"weihanglo","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/3545958152/jobs/5954969068"},"merge_sha":"de56c1251bfe8f864112bc6cd9e9d73fee71fac4"} -->
